build: migration to java 21, spring boot 3.5

cabbage16 opened the pull request:
## 🎫 관련 이슈
[//]: # (다음 키워드를 사용하면 해당 PR을 머지할 때 자동으로 이슈를 닫을 수 있습니다.)
[//]: # (keyword: close|closes|closed|resolve|resolves|resolved|fix|fixes|fixed)
[//]: # (예시: close #1)

#321 

<br>

## 📄 개요
[//]: # (작업 내용을 간단히 요약해서 적습니다.)
[//]: # (예시: 유저 회원가입 기능을 만들었습니다.)

> JVM과 Spring Boot, 기타 라이브러리들의 버전을 업그레이드하고 그에 따른 에러를 해결했습니다.

<br>

## 🔨 작업 내용
[//]: # (작업 내용을 자세하게 적습니다.)
[//]: # (붙임표 "-" 을 사용해서 목록을 만듭니다.)
[//]: # (예시: 유저 회원가입 API를 만들었습니다.)

[Spring Boot 4.0 Migration Guide](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-4.0-Migration-Guide)에 따라 4.0으로 마이그레이션 하기 전 우선 3.5.X 버전으로 업그레이드 하는 PR입니다.
3.5.8 업그레이드에 성공하면 해당 가이드에 따라 Java 25, Spring Boot 4.0 까지 업그레이드 할 계획입니다.
- Java 17 -> 21
- Spring Boot 3.0.5 -> 3.5.8
- `NoResourceFoundException` 핸들링(Spring 6.1부터 더이상 직접 404를 반환하지 않고 해당 에러를 발생시킴)
- 기타 라이브러리 버전업 및 지원 중단된 문법 수정


<br>

## 🏁 확인 사항
[//]: # (PR을 보내기 전 다음 사항을 확인해주세요.)
[//]: # (해당 사항을 모두 이행해야 머지할 수 있습니다.)
[//]: # (- [x] 를 사용해서 완료로 표시할 수 있습니다.)

- [x]  테스트를 완료했나요?
- [ ] API 문서를 작성했나요?
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 로그, 주석, import 등을 삭제했나요?

<br>

## 🙋🏻 덧붙일 말
[//]: # (다음 사항이 있다면 적어주세요.)
[//]: # (PR에 대한 추가 설명)
[//]: # (중점적으로 리뷰받고 싶은 부분)
[//]: # (기타 등등)

@Bamdoliro/maru-backend 각자 환경에서도 한번씩 실행시켜봐주시고 문제 발생하면 comment 부탁드립니다!

cabbage16 committed:
chore: build.gradle 의존성 그룹화 및 정리

cabbage16 committed:
build: Gradle 7.6.1 -> 8.14

cabbage16 committed:
build: Java, Spring Boot, 기타 라이브러리 버전업

- Java 17 -> 21
- Spring Boot 3.0.5 -> 3.5.8
- Spring Dependency Management 1.1.0 -> 1.1.7
- Spring Cloud 2022.0.2 -> 2025.0.0

cabbage16 committed:
fix: Deprecated된 HttpSecurity문법 수정

cabbage16 committed:
Merge branch 'Bamdoliro:develop' into build/#321

cabbage16 committed:
fix: Not Found 에러 핸들링

- 상위버전에서는 404를 자동으로 반환하지 않고 NoResourceFoundException을 발생시키므로 이에 대한 핸들링 추가

cabbage16 committed:
test: NoticeControllerTest 오류 해결

- 빈 FileMetadata로 인한 오류 해결

cabbage16 committed:
chore: unused import

cabbage16 committed:
ci: upgrade to java 21

cabbage16 commented:
`@MockBean` deprecated 👀

jyj1289 reviewed (approved):
@cabbage16  Sonar Cloud 터지는 것만 확인해주세요

cabbage16 commented:
> @cabbage16 Sonar Cloud 터지는 것만 확인해주세요

@jyj1289 CSRF disable 때문인데 기존 코드에서도 CSRF disable 하고있었습니다. 아니면 어차피 지금 마루의 인증/인가 로직에서 스프링 시큐리티를 전혀 사용하고있는 것 같지 않은데 아예 spring security 의존성을 빼는것도 고려해볼만 한 것 같습니다

cabbage16 committed:
fix: @MockBean의 지원 중단으로이해 @MockitoBean 으로 변경

cabbage16 committed:
build: build.gradle 문법 수정

- sourceCompatibility -> JavaLanguageVersion
- mockito-inline -> mockito-core
- $buildDir -> layout.buildDirectory.dir()
